### PR TITLE
Bugfix/websocket authorization

### PR DESCRIPTION
**Pull request by nikkiki**
## What is this change?

Adds basic authorization middleware to the agent web socket connection.

## Why is this change necessary?

Closes https://github.com/sensu/sensu-go/issues/2811

## Does your change need a Changelog entry?

Yas.

## Do you need clarification on anything?

Nah.

## Were there any complications while making this change?

Yes, needed to pair with @palourde and @ccressent for a bit, but they were able to clarify the missing piece for me! (it was injecting the namespace into the context)

## Have you reviewed and updated the documentation for this change? Is new documentation required?

It's a bugfix, docs should be good.

## How did you verify this change?

Integration and e2e testing. New test case is probably not required.

### CODE REVIEW

File: backend/agentd/agentd.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ type Agentd struct {
 	wg         *sync.WaitGroup
 	errChan    chan error
 	httpServer *http.Server
-	store      Store
+	store      store.Store
```
Comment (palourde) [inline backend/agentd/agentd.go]: Can we remove the Store interface at line 28?

File: backend/apid/middlewares/authorization.go [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
 import (
+	"context"
 	"net/http"
 
 	"github.com/sensu/sensu-go/backend/apid/actions"
 	"github.com/sensu/sensu-go/backend/authorization"
+	"github.com/sensu/sensu-go/backend/authorization/rbac"
+	"github.com/sensu/sensu-go/backend/store"
+	"github.com/sensu/sensu-go/transport"
+	corev2 "github.com/sensu/sensu-go/types"
```
Comment (palourde) [inline backend/apid/middlewares/authorization.go]: Could we replace `github.com/sensu/sensu-go/types` with `github.com/sensu/sensu-go/api/core/v2`?

File: backend/apid/middlewares/authorization.go [hunk 3]
```diff
@@ -44,3 +49,41 @@ func (a Authorization) Then(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r.WithContext(ctx))
 	})
 }
+
+// BasicAuthorization performs basic authorization for entity creation via the agent websocket.
```
Comment (palourde) [inline backend/apid/middlewares/authorization.go]: s/entity/event/

File: backend/apid/middlewares/authorization.go [hunk 4]
```diff
@@ -44,3 +49,41 @@ func (a Authorization) Then(next http.Handler) http.Handler {
...
+func BasicAuthorization(next http.Handler, store store.Store) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		namespace := r.Header.Get(transport.HeaderKeyNamespace)
+		ctx := r.Context()
+		ctx = context.WithValue(ctx, corev2.NamespaceKey, namespace)
+
+		user, err := store.GetUser(ctx, r.Header.Get(transport.HeaderKeyUser))
```
Comment (palourde) [inline backend/apid/middlewares/authorization.go]: I don't believe we should do this, considering the agent user could be within an authentication provider. Instead, we should rely on the claims injected into the context by [`middlewares.BasicAuthentication`](https://github.com/sensu/sensu-go/blob/c80ddead34075326a45ae9d3f878304c6c43ce53/backend/apid/middlewares/authentication.go#L84) and fetch those like this: https://github.com/sensu/sensu-go/blob/a82a17edaf9f7455c5d86cc2a63bbe5bfc152835/backend/apid/middlewares/authorization_attributes.go#L120

Comment (nikkiki) [reply backend/apid/middlewares/authorization.go]: Good point! Fixing this also made me realize I had the authorization and authentication middlewares in the wrong order. 2 🐦1 stone!

**Pull request merged**